File: prototype/csharp_interop_prototype/cpp_files/src/component.cpp
//
// Created by charles on 08/11/2025.
//

#include "component.h"

component::component(void *_gc_handle)
    : m_gc_handle(_gc_handle)
{}

component::~component()
{
    if (g_destroy_ptr)
    {
        g_destroy_ptr(m_gc_handle);
    }
}

void component::start()
{
    if (g_start_ptr)
    {
        g_start_ptr(m_gc_handle);
    }
}

void component::update()
{
    if (g_update_ptr)
    {
        g_update_ptr(m_gc_handle);
    }
}

void set_global_component_callbacks(start_ptr_t _start_ptr, update_ptr_t _update_ptr, destroy_ptr_t _destroy_ptr)
{
    g_start_ptr = _start_ptr;
    g_update_ptr = _update_ptr;
    g_destroy_ptr = _destroy_ptr;
}

File: prototype/csharp_interop_prototype/csharp_files/Component.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

namespace interop_testing;

internal static partial class ExternalComponent
{
    [LibraryImport(libraryName: "InteropCPP", EntryPoint = "set_global_component_callbacks", StringMarshalling = StringMarshalling.Utf8)]
    [UnmanagedCallConv(CallConvs = new []{typeof(CallConvCdecl)})]
    internal static unsafe partial void SetGlobalComponentCallback(
        delegate* unmanaged[Cdecl]<IntPtr, void> startPtr, 
        delegate* unmanaged[Cdecl]<IntPtr, void> updatePtr,
        delegate* unmanaged[Cdecl]<IntPtr, void> destroyPtr);
    
    [UnmanagedCallersOnly(CallConvs = new []{typeof(CallConvCdecl)})]
    internal static void ComponentStart(IntPtr gcHandlePtr)
    {
        GCHandle gcHandle = GCHandle.FromIntPtr(gcHandlePtr);
        Component component = (Component)gcHandle.Target;
        component?.Start();
    }
    
    [UnmanagedCallersOnly(CallConvs = new []{typeof(CallConvCdecl)})]
    internal static void ComponentUpdate(IntPtr gcHandlePtr)
    {
        GCHandle gcHandle = GCHandle.FromIntPtr(gcHandlePtr);
        Component component = (Component)gcHandle.Target;
        component?.Update();
    }

    [UnmanagedCallersOnly(CallConvs = new []{typeof(CallConvCdecl)})]
    internal static void ComponentDestroy(IntPtr gcHandlePtr)
    {
        GCHandle gcHandle = GCHandle.FromIntPtr(gcHandlePtr);
        ((Component)gcHandle.Target)?.Destroy();
        gcHandle.Free();
    }
}

public class Component
{
    internal IntPtr InternalReference;

    public virtual void Start() {}
    
    public virtual void Update() {}
    
    public virtual void Destroy() {}
}